# Machine Learning Project - Kickstarter Data Set
## Data Cleaning Notebook
*Contributor: Max Langer, René Ebrecht, Jens Reich*

This is the very first project where we build a machine learning model from scratch based on an unknown dataset.
The dataset includes data from Kickstarter projects from the years 2009 to 2019.

Our goal is to help our (fictional) stackholder, PPC Consultants with a model that can predict whether a Kickstarter project will be successful or not. 
PPC Consultants advises potential project creators (PPCs) with their projects to get them off the ground as successfully as possible.
Therefore, the value of our data product (the predictive model) is to show opportunities, save time, and in the end make money for both PPC consultants and PPCs.

In [1]:
# Import the organization modules
import pandas as pd
import numpy as np
# Import module to ignore warnings
import warnings
warnings.filterwarnings('ignore')
# Import the plot modules
import matplotlib.pyplot as plt
import seaborn as sns
# Import own scripts
from scripts.data_cleaning import (
    read_all_csvs, 
    data_cleaning,
    create_csv, 
    get_nan_cols, 
    convert_to_datetime, 
    calculate_time_periods, 
    get_year_month_day, 
    entangle_column
    )

In [2]:
# Create data frame from all single CSV files
df = read_all_csvs()

In [3]:
# Take a look at the first columns
df.head()

backers_count                                              blurb  \
0             21  2006 was almost 7 years ago.... Can you believ...   
1             97  An adorable fantasy enamel pin series of princ...   
2             88  Helping a community come together to set the s...   
3            193  Every revolution starts from the bottom and we...   
4             20  Learn to build 10+ Applications in this comple...   

                                            category  \
0  {"id":43,"name":"Rock","slug":"music/rock","po...   
1  {"id":54,"name":"Mixed Media","slug":"art/mixe...   
2  {"id":280,"name":"Photobooks","slug":"photogra...   
3  {"id":266,"name":"Footwear","slug":"fashion/fo...   
4  {"id":51,"name":"Software","slug":"technology/...   

   converted_pledged_amount country  created_at  \
0                       802      US  1387659690   
1                      2259      US  1549659768   
2                     29638      US  1477242384   
3                     49158      IT  1540369920   
4                       549      US  1425706517   

                                             creator currency currency_symbol  \
0  {"id":1495925645,"name":"Daniel","is_registere...      USD               $   
1  {"id":1175589980,"name":"Katherine","slug":"fr...      USD               $   
2  {"id":1196856269,"name":"MelissaThomas","is_re...      USD               $   
3  {"id":1569700626,"name":"WAO","slug":"wearewao...      EUR               €   
4  {"id":1870845385,"name":"Kalpit Jain","is_regi...      USD               $   

   currency_trailing_code  ...  \
0                    True  ...   
1                    True  ...   
2                    True  ...   
3                   False  ...   
4                    True  ...   

                                                slug  \
0                              new-final-round-album   
1                    princess-pals-enamel-pin-series   
2  their-life-through-their-lens-the-amish-and-me...   
3                           wao-the-eco-effect-shoes   
4                     apple-watch-development-course   

                                          source_url  spotlight staff_pick  \
0  https://www.kickstarter.com/discover/categorie...       True      False   
1  https://www.kickstarter.com/discover/categorie...       True      False   
2  https://www.kickstarter.com/discover/categorie...       True       True   
3  https://www.kickstarter.com/discover/categorie...       True      False   
4  https://www.kickstarter.com/discover/categorie...      False      False   

        state  state_changed_at  static_usd_rate  \
0  successful        1391899046         1.000000   
1  successful        1551801611         1.000000   
2  successful        1480607932         1.000000   
3  successful        1544309940         1.136525   
4      failed        1428511019         1.000000   

                                                urls  usd_pledged  \
0  {"web":{"project":"https://www.kickstarter.com...    802.00000   
1  {"web":{"project":"https://www.kickstarter.com...   2259.00000   
2  {"web":{"project":"https://www.kickstarter.com...  29638.00000   
3  {"web":{"project":"https://www.kickstarter.com...  49075.15252   
4  {"web":{"project":"https://www.kickstarter.com...    549.00000   

        usd_type  
0  international  
1  international  
2  international  
3  international  
4       domestic  

[5 rows x 37 columns]

In [4]:
# Prints out the columns with NaNs
get_nan_cols(df)

Number of NaNs per column:
blurb:8
friends:208922
is_backing:208922
is_starred:208922
location:226
permissions:208922
usd_type:480


We can get rid of some columns, which probably will have no impact on the model later on. 
Especially `friends`, `is_backing`, `is_starred`, `permission` are almost completely empty (over 50% of data was missing). So no information loss here.

Moreover the following columns don't seem to contain any valuable information for the predictions as well: `blurb`,`converted_pledged_amount`,`currency_symbol`, `currency_trailing_code`, `fx_rate`, `photo`,`profile`,`slug`,`source_url`,`static_usd_rate`,`urls`.

The 226 and 480 missing observations for `location` and `usd_type` are so small in number that we can get rid of these rows as well. 

We also rename the `currency` into `original_currency`, and `id` int `project_id`, as this gives more information about its content. 

In [5]:
# Convert the column names to a list
columns = df.columns.to_list()
# Drop all columns with more than 50% of the observations missing
df = df[[column for column in df if df[column].count() / len(df) >= 0.5]]
# Print the dropped columns
print("Dropped columnes:", end= " ")
dropped = [print(col, end=" ") for col in columns if col not in df.columns]

Dropped columnes: friends is_backing is_starred permissions 

In [6]:
# Drop the listed columns
df.drop([
    'blurb', 
    'converted_pledged_amount',
    'currency_symbol', 
    'currency_trailing_code', 
    'fx_rate',
    'photo',
    'profile',
    'slug',
    'source_url', 
    'static_usd_rate',
    'urls'
    ], axis=1, inplace=True)
# Drops the last few NaN values
df.dropna(axis=0, inplace=True)
# Rename the currency column
df.rename(columns={
    'currency':'original_currency',
    'id':'project_id'}, inplace=True)

In [7]:
# Prints out the columns with NaNs
get_nan_cols(df)

Number of NaNs per column:


Now we have no NaNs left in our data set, which is another step towards a cleaned data set.

The columns of `created_at`, `state_changed_at`, `deadline` are actually dates but expressed as epoch time. We will convert these into datetime objects and thereby into actual dates.

With these we can calculate the time periods from the beginning of the project until the success and also to the deadline of the project.

Moreover, from these dates we can then get the year, month and day of the three columns.

In [8]:
# Convert the time columns to datetime types
df = convert_to_datetime(df, ['created_at', 'state_changed_at', 'deadline'])
# Calculate the time periods
df = calculate_time_periods(df)
# Get the years, months and days as separate columns
df = get_year_month_day(df, ['created_at', 'state_changed_at', 'deadline'])

In [9]:
df.head()

backers_count                                           category country  \
0             21  {"id":43,"name":"Rock","slug":"music/rock","po...      US   
1             97  {"id":54,"name":"Mixed Media","slug":"art/mixe...      US   
2             88  {"id":280,"name":"Photobooks","slug":"photogra...      US   
3            193  {"id":266,"name":"Footwear","slug":"fashion/fo...      IT   
4             20  {"id":51,"name":"Software","slug":"technology/...      US   

                                             creator original_currency  \
0  {"id":1495925645,"name":"Daniel","is_registere...               USD   
1  {"id":1175589980,"name":"Katherine","slug":"fr...               USD   
2  {"id":1196856269,"name":"MelissaThomas","is_re...               USD   
3  {"id":1569700626,"name":"WAO","slug":"wearewao...               EUR   
4  {"id":1870845385,"name":"Kalpit Jain","is_regi...               USD   

  current_currency  disable_communication     goal  project_id  is_starrable  \
0              USD                  False    200.0   287514992         False   
1              USD                  False    400.0   385129759         False   
2              USD                  False  27224.0   681033598         False   
3              USD                  False  40000.0  1031782682         False   
4              USD                  False   1000.0   904085819         False   

   ...  days_total created_at_year created_at_month  created_at_day  \
0  ...          49            2013               12              21   
1  ...          25            2019                2               8   
2  ...          39            2016               10              23   
3  ...          45            2018               10              24   
4  ...          32            2015                3               7   

   state_changed_at_year  state_changed_at_month state_changed_at_day  \
0                   2014                       2                    8   
1                   2019                       3                    5   
2                   2016                      12                    1   
3                   2018                      12                    8   
4                   2015                       4                    8   

   deadline_year deadline_month  deadline_day  
0           2014              2             8  
1           2019              3             5  
2           2016             12             1  
3           2018             12             8  
4           2015              4             8  

[5 rows x 30 columns]

In [10]:
print("Percentage of the data where the project duration is the same as the time periode to reach the goal:")
print((df[df['days_till_change']==df['days_total']].shape[0] / len(df))*100)

Percentage of the data where the project duration is the same as the time periode to reach the goal:
92.39243031709798


We see that 92% of the projects reach the goal in the same number of days as the deadline. Perhaps they were simply marked as completed after the target was reached. 

Let's look further into the data types of the other columns.

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 208516 entries, 0 to 964
Data columns (total 30 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   backers_count           208516 non-null  int64  
 1   category                208516 non-null  object 
 2   country                 208516 non-null  object 
 3   creator                 208516 non-null  object 
 4   original_currency       208516 non-null  object 
 5   current_currency        208516 non-null  object 
 6   disable_communication   208516 non-null  bool   
 7   goal                    208516 non-null  float64
 8   project_id              208516 non-null  int64  
 9   is_starrable            208516 non-null  bool   
 10  launched_at             208516 non-null  int64  
 11  location                208516 non-null  object 
 12  name                    208516 non-null  object 
 13  pledged                 208516 non-null  float64
 14  spotlight              

In [12]:
for col in ['category', 'creator', 'location']:
    print(df[col].iloc[0], '\n')

{"id":43,"name":"Rock","slug":"music/rock","position":17,"parent_id":14,"color":10878931,"urls":{"web":{"discover":"http://www.kickstarter.com/discover/categories/music/rock"}}} 

{"id":1495925645,"name":"Daniel","is_registered":null,"chosen_currency":null,"avatar":{"thumb":"https://ksr-ugc.imgix.net/assets/006/041/047/c44d1a95c2139ae46af635c7c6e7ea76_original.jpg?ixlib=rb-1.1.0&w=40&h=40&fit=crop&v=1461362658&auto=format&frame=1&q=92&s=3d655afafac9dbb59c1e675adfa87082","small":"https://ksr-ugc.imgix.net/assets/006/041/047/c44d1a95c2139ae46af635c7c6e7ea76_original.jpg?ixlib=rb-1.1.0&w=160&h=160&fit=crop&v=1461362658&auto=format&frame=1&q=92&s=3973d24f5c3db1ed1d5c84cec8af1d6d","medium":"https://ksr-ugc.imgix.net/assets/006/041/047/c44d1a95c2139ae46af635c7c6e7ea76_original.jpg?ixlib=rb-1.1.0&w=160&h=160&fit=crop&v=1461362658&auto=format&frame=1&q=92&s=3973d24f5c3db1ed1d5c84cec8af1d6d"},"urls":{"web":{"user":"https://www.kickstarter.com/profile/1495925645"},"api":{"user":"https://api.kick

We find very strange formatted object type data in `category`, `creator` and `location` column. The next step will be to entangle these strings and get the most important information out of them. For the location we could extract the city name of the project creator. For the creator we could get their names for now. And for the category the id and the name could be interesting.

In [13]:
# Entangles the category, creator and location column
df = entangle_column(df, ['category', 'creator', 'location'])

In [14]:
df.head()

backers_count country original_currency current_currency  \
0             21      US               USD              USD   
1             97      US               USD              USD   
2             88      US               USD              USD   
3            193      IT               EUR              USD   
4             20      US               USD              USD   

   disable_communication     goal  project_id  is_starrable  launched_at  \
0                  False    200.0   287514992         False   1388011046   
1                  False    400.0   385129759         False   1550073611   
2                  False  27224.0   681033598         False   1478012330   
3                  False  40000.0  1031782682         False   1540684582   
4                  False   1000.0   904085819         False   1425919017   

                                                name  ...  \
0                              New Final Round Album  ...   
1                    Princess Pals Enamel Pin Series  ...   
2  Their Life Through Their Lens-the Amish and Me...  ...   
3                          WAO: THE ECO EFFECT SHOES  ...   
4                     Apple Watch Development Course  ...   

   state_changed_at_month  state_changed_at_day  deadline_year deadline_month  \
0                       2                     8           2014              2   
1                       3                     5           2019              3   
2                      12                     1           2016             12   
3                      12                     8           2018             12   
4                       4                     8           2015              4   

   deadline_day category_name  category_id  category_sub   creator_name  \
0             8         music           43          rock         Daniel   
1             5           art           54   mixed media      Katherine   
2             1   photography          280    photobooks  MelissaThomas   
3             8       fashion          266      footwear            WAO   
4             8    technology           51      software    Kalpit Jain   

   location_name  
0        Chicago  
1     Sacramento  
2       Columbus  
3         Venice  
4        Redmond  

[5 rows x 32 columns]

In [15]:
get_nan_cols(df)

Number of NaNs per column:


Now we should take a look at the number of unique observations for each column.

In [16]:
# Show the amount of unique items per column
df.nunique()

backers_count               3237
country                       22
original_currency             14
current_currency               1
disable_communication          2
goal                        5103
project_id                182004
is_starrable                   2
launched_at               181849
name                      181421
pledged                    44293
spotlight                      2
staff_pick                     2
state                          5
usd_pledged                79114
usd_type                       2
days_till_change            1501
days_total                  1497
created_at_year               11
created_at_month              12
created_at_day                31
state_changed_at_year         11
state_changed_at_month        12
state_changed_at_day          31
deadline_year                 11
deadline_month                12
deadline_day                  31
category_name                 15
category_id                  169
category_sub                 145
creator_na

If we look at the number of unique entries for each column, we can see that the current currency has only one entry, USD. Since this does not add any information to the model, we can also drop this column. 

In [17]:
# Drop the current_currency column
df.drop('current_currency', axis=1, inplace=True)

In [18]:
df.head()

backers_count country original_currency  disable_communication     goal  \
0             21      US               USD                  False    200.0   
1             97      US               USD                  False    400.0   
2             88      US               USD                  False  27224.0   
3            193      IT               EUR                  False  40000.0   
4             20      US               USD                  False   1000.0   

   project_id  is_starrable  launched_at  \
0   287514992         False   1388011046   
1   385129759         False   1550073611   
2   681033598         False   1478012330   
3  1031782682         False   1540684582   
4   904085819         False   1425919017   

                                                name  pledged  ...  \
0                              New Final Round Album    802.0  ...   
1                    Princess Pals Enamel Pin Series   2259.0  ...   
2  Their Life Through Their Lens-the Amish and Me...  29638.0  ...   
3                          WAO: THE ECO EFFECT SHOES  43180.0  ...   
4                     Apple Watch Development Course    549.0  ...   

   state_changed_at_month  state_changed_at_day deadline_year  deadline_month  \
0                       2                     8          2014               2   
1                       3                     5          2019               3   
2                      12                     1          2016              12   
3                      12                     8          2018              12   
4                       4                     8          2015               4   

  deadline_day  category_name  category_id  category_sub   creator_name  \
0            8          music           43          rock         Daniel   
1            5            art           54   mixed media      Katherine   
2            1    photography          280    photobooks  MelissaThomas   
3            8        fashion          266      footwear            WAO   
4            8     technology           51      software    Kalpit Jain   

   location_name  
0        Chicago  
1     Sacramento  
2       Columbus  
3         Venice  
4        Redmond  

[5 rows x 31 columns]

In [19]:
# Create a list of the real numeric columns
numeric_cols = [
    'backers_count', 
    'goal', 
    'pledged',
    'usd_pledged',
    'days_till_change',
    'days_total']
# Get the basic descriptive statistics for the numeric columns
df[numeric_cols].describe()

backers_count          goal       pledged   usd_pledged  \
count  208516.000000  2.085160e+05  2.085160e+05  2.085160e+05   
mean      145.277734  4.926378e+04  1.881146e+04  1.289545e+04   
std       884.468471  1.181398e+06  3.234143e+05  8.896665e+04   
min         0.000000  1.000000e-02  0.000000e+00  0.000000e+00   
25%         4.000000  1.500000e+03  1.090000e+02  1.060000e+02   
50%        27.000000  5.000000e+03  1.555000e+03  1.537000e+03   
75%        89.000000  1.500000e+04  6.889057e+03  6.552000e+03   
max    105857.000000  1.000000e+08  8.103074e+07  8.596475e+06   

       days_till_change     days_total  
count     208516.000000  208516.000000  
mean          78.145221      80.340765  
std          132.063629     131.917357  
min            0.000000       1.000000  
25%           32.000000      33.000000  
50%           45.000000      47.000000  
75%           71.000000      73.000000  
max         3334.000000    3334.000000

In [21]:
df.groupby(['category_id', 'category_name']).count()

backers_count  country  original_currency  \
category_id category_name                                              
1           art                     1019     1019               1019   
10          food                     568      568                568   
11          film & video             276      276                276   
12          games                     95       95                 95   
13          journalism               699      699                699   
...                                  ...      ...                ...   
53          art                     2496     2496               2496   
54          art                     2546     2546               2546   
6           dance                   1478     1478               1478   
7           design                   249      249                249   
9           fashion                  194      194                194   

                           disable_communication  goal  project_id  \
category_id category_name                                            
1           art                             1019  1019        1019   
10          food                             568   568         568   
11          film & video                     276   276         276   
12          games                             95    95          95   
13          journalism                       699   699         699   
...                                          ...   ...         ...   
53          art                             2496  2496        2496   
54          art                             2546  2546        2546   
6           dance                           1478  1478        1478   
7           design                           249   249         249   
9           fashion                          194   194         194   

                           is_starrable  launched_at  name  pledged  ...  \
category_id category_name                                            ...   
1           art                    1019         1019  1019     1019  ...   
10          food                    568          568   568      568  ...   
11          film & video            276          276   276      276  ...   
12          games                    95           95    95       95  ...   
13          journalism              699          699   699      699  ...   
...                                 ...          ...   ...      ...  ...   
53          art                    2496         2496  2496     2496  ...   
54          art                    2546         2546  2546     2546  ...   
6           dance                  1478         1478  1478     1478  ...   
7           design                  249          249   249      249  ...   
9           fashion                 194          194   194      194  ...   

                           created_at_day  state_changed_at_year  \
category_id category_name                                          
1           art                      1019                   1019   
10          food                      568                    568   
11          film & video              276                    276   
12          games                      95                     95   
13          journalism                699                    699   
...                                   ...                    ...   
53          art                      2496                   2496   
54          art                      2546                   2546   
6           dance                    1478                   1478   
7           design                    249                    249   
9           fashion                   194                    194   

                           state_changed_at_month  state_changed_at_day  \
category_id category_name                                                 
1           art                              1019                  1019   
10          food                              568                   568   
11   

In [4]:
df.head()

backers_count country original_currency  disable_communication     goal  \
0             21      US               USD                  False    200.0   
1             97      US               USD                  False    400.0   
2             88      US               USD                  False  27224.0   
3            193      IT               EUR                  False  40000.0   
4             20      US               USD                  False   1000.0   

   project_id  is_starrable  launched_at  \
0   287514992         False   1388011046   
1   385129759         False   1550073611   
2   681033598         False   1478012330   
3  1031782682         False   1540684582   
4   904085819         False   1425919017   

                                                name  pledged  ...  \
0                              New Final Round Album    802.0  ...   
1                    Princess Pals Enamel Pin Series   2259.0  ...   
2  Their Life Through Their Lens-the Amish and Me...  29638.0  ...   
3                          WAO: THE ECO EFFECT SHOES  43180.0  ...   
4                     Apple Watch Development Course    549.0  ...   

   state_changed_at_month  state_changed_at_day deadline_year  deadline_month  \
0                       2                     8          2014               2   
1                       3                     5          2019               3   
2                      12                     1          2016              12   
3                      12                     8          2018              12   
4                       4                     8          2015               4   

  deadline_day  category_name  category_id  category_sub   creator_name  \
0            8          music           43          rock         Daniel   
1            5            art           54   mixed media      Katherine   
2            1    photography          280    photobooks  MelissaThomas   
3            8        fashion          266      footwear            WAO   
4            8     technology           51      software    Kalpit Jain   

   location_name  
0        Chicago  
1     Sacramento  
2       Columbus  
3         Venice  
4        Redmond  

[5 rows x 31 columns]